src/litdata/utilities/shuffle.py

```diff
@@ -71,18 +71,31 @@ def _associate_chunks_and_intervals_to_workers(
     batch_size: int = 1,
 ) -> Tuple[List[List[int]], List[Any]]:
     num_items = sum([(interval[2] - interval[1]) for interval in chunk_intervals])
-    world_size = distributed_env.world_size * num_workers
-    num_items_per_workers: List[int] = [
-        num_items // world_size + num_items % world_size
-        if rank == world_size - 1 and not drop_last
-        else num_items // world_size
-        for rank in range(world_size)
-    ]
-    if drop_last:
-        num_items_per_workers = [batch_size * int(item // batch_size) for item in num_items_per_workers]
-
-    chunks_per_workers: List[List[int]] = [[] for _ in range(world_size)]
-    intervals_per_workers: List[List[List[int]]] = [[] for _ in range(world_size)]
+    max_batches = num_items // batch_size
+    global_num_workers = distributed_env.world_size * num_workers
+
+    num_batches_per_workers = []
+
+    for rank in range(distributed_env.world_size):
+        tmp_arr = [0 for _ in range(num_workers)]
+
+        index = 0
+        for _ in range(int(max_batches // distributed_env.world_size)):
+            tmp_arr[index % num_workers] += 1
+            index += 1
+
+        if not drop_last:
+            tmp_arr[index % num_workers] += (num_items - max_batches * batch_size) // distributed_env.world_size
```

Comment:
this logic can fail if the remainder is less than distributed_env.world_size. For instance:
- Suppose remainder = 2, but distributed_env.world_size = 4.
- Then (remainder) // 4 = 0, so each of the first three ranks get +0.
- On the last rank: 2 - 4 = -2, which is negative, leading to effectively dropping items (or messing up the distribution).

Reply:
I changed the logic.